feat(webdav): add bearer_token_command for dynamic token acquisition

GreenHatHG opened the pull request:
see: https://github.com/iterative/dvc/pull/10917

GreenHatHG committed:
feat(webdav): add bearer_token_command config for dynamic token retrieval

GreenHatHG committed:
refactor(dvc_webdav): ensure BearerAuthClient is initialized only once

GreenHatHG committed:
refactor: fix linter warnings

GreenHatHG committed:
refactor(webdav): remove token persistence logic to enforce in-memory usage

skshetry commented on dvc_webdav/bearer_auth_client.py:
```diff
@@ -0,0 +1,161 @@
+    return token
+
+
+class BearerAuthClient(httpx.Client):
```
Do we need to extend `httpx.Client`? Could we make this work with custom [`httpx.Auth`](https://www.python-httpx.org/advanced/authentication/)?

GreenHatHG committed:
feat(webdav): refactor Bearer auth to use httpx.Auth with automatic 401 retry

skshetry reviewed (approved):
Thank you for contributing. This looks good overall. One question. Could you also add a few tests to exercise the `BearerAuth` code path? That would help ensure it works correctly and does not regress.

Also, contribution to dvc repositories now require accepting the CLA: https://cla-assistant.io/treeverse/dvc-webdav. It'd be great if you could do that. Thanks.

GreenHatHG committed:
fix(dvc_webdav): remove manual response body reading in BearerAuth

GreenHatHG committed:
tests(BearerAuth): add test cases

GreenHatHG commented on dvc_webdav/bearer_auth.py:
```diff
@@ -0,0 +1,159 @@
+            try:
+                response.read()
+            except Exception as e:  # noqa: BLE001
+                logger.debug("Error reading response body: %s", e)
```
Fixed in 773944a.
I verified that removing `response.read()` works correctly.

GreenHatHG commented:
@skshetry 
I've added the test cases covering `BearerAuth` as requested.
Also, to ensure end-to-end stability, I've added a corresponding integration test in the dvc repo (treeverse/dvc#10917).

skshetry commented on dvc_webdav/tests/test_bearer_auth.py:
```diff
@@ -0,0 +1,329 @@
+    # indicating no further retries will be attempted.
+    with pytest.raises(StopIteration):
+        flow.send(resp_401_again)
+
```
I don’t see the need for tests beyond this line, especially the concurrent one. I doubt they would help catch the issue if it occurs in the future.

When I requested tests, my only concern was verifying that the command is called and behaves as expected: setting the token in the header on success and raising an error when it exits with a failure. Another goal was to ensure it works on Windows, though the current test doesn’t cover that since all subprocess calls are mocked and I’m okay with that.

I appreciate that you also added tests for token refresh, but the rest feels excessive.

GreenHatHG replied:
Thanks for the feedback!
I agree that the test suite was getting a bit heavy. I've removed `test_auth_flow_failure_after_refresh` and the other edge-case tests as suggested.
It looks much cleaner now.  Please let me know if there's anything else needed.

GreenHatHG committed:
tests(BearerAuth): remove excessive bearer_auth test cases

GreenHatHG commented:
@skshetry 
It seems the GitHub Actions workflow is waiting for approval to run. Could you please hit the "Approve and run" button when you have a moment? Thanks!
<img width="654" height="273" alt="图片" src="https://github.com/user-attachments/assets/a9bc00b7-4e1b-4e40-b9ad-a48cbc0ae240" />

GreenHatHG committed:
test: add pytest-mock as test dependency

GreenHatHG commented:
> @GreenHatHG, looks like we don't have `pytest-mock` installed here. Please feel add them to `tests` dependencies in pyproject.toml.

Fixed in the latest commit. 
All CI checks have passed. The remaining conversation threads are addressed in the code.
I think this is ready to be merged whenever you have a moment.
